tuf/ngclient/updater.py

```diff
@@ -0,0 +1,430 @@
+            logger.debug("Local snapshot is valid: not downloading new one")
+        except (OSError, exceptions.RepositoryError) as e:
+            # Local snapshot does not exist or is invalid: update from remote
+            logger.debug("Failed to load local snapshot %s", e)
+
+            metainfo = self._bundle.timestamp.signed.meta["snapshot.json"]
+            length = metainfo.length or DEFAULT_SNAPSHOT_MAX_LENGTH
+            version = None
+            if self._bundle.root.signed.consistent_snapshot:
+                version = metainfo.version
+
+            data = self._download_metadata("snapshot", length, version)
+            self._bundle.update_snapshot(data)
+            self._persist_metadata("snapshot", data)
+
+    def _load_targets(self, role: str, parent_role: str) -> None:
+        """Load local (and if needed remote) metadata for 'role'."""
+        try:
+            data = self._load_local_metadata(role)
+            self._bundle.update_delegated_targets(data, role, parent_role)
+            logger.debug("Local %s is valid: not downloading new one", role)
+        except (OSError, exceptions.RepositoryError) as e:
+            # Local 'role' does not exist or is invalid: update from remote
+            logger.debug("Failed to load local %s: %s", role, e)
+
+            metainfo = self._bundle.snapshot.signed.meta[f"{role}.json"]
+            length = metainfo.length or DEFAULT_TARGETS_MAX_LENGTH
+            version = None
+            if self._bundle.root.signed.consistent_snapshot:
+                version = metainfo.version
+
+            data = self._download_metadata(role, length, version)
+            self._bundle.update_delegated_targets(data, role, parent_role)
+            self._persist_metadata(role, data)
+
+    def _preorder_depth_first_walk(self, target_filepath: str) -> Dict:
+        """
+        Interrogates the tree of target delegations in order of appearance
+        (which implicitly order trustworthiness), and returns the matching
+        target found in the most trusted role.
+        """
+
+        role_names = [("targets", "root")]
+        visited_role_names = set()
+        number_of_delegations = MAX_DELEGATIONS
+
+        # Preorder depth-first traversal of the graph of target delegations.
+        while number_of_delegations > 0 and len(role_names) > 0:
+
+            # Pop the role name from the top of the stack.
+            role_name, parent_role = role_names.pop(-1)
+
+            # Skip any visited current role to prevent cycles.
+            if (role_name, parent_role) in visited_role_names:
+                logger.debug("Skipping visited current role %s", role_name)
+                continue
+
+            # The metadata for 'role_name' must be downloaded/updated before
+            # its targets, delegations, and child roles can be inspected.
+            self._load_targets(role_name, parent_role)
+
+            role_metadata = self._bundle[role_name].signed
```

Comment:
this implies we should make bundle (or trusted_set as it maybe now) return the loaded metadata to avoid a lookup immediately afterwards. Can you file an issue please?

Reply:
Done #1507